runtime/stdlib.js

```diff
@@ -423,8 +424,9 @@ function caml_compare_val (a, b, total) {
             var comp = caml_compare_val_get_custom(a);
             if(comp) {
               var x = comp(a,b,total);
-              if(!(total && x != x) || (x === (x|0) && x != 0))
-                return x
+              if(total && x != x) return 1;
+              if(+x != +x) return +x;
+              if((x | 0) != 0) return (x | 0);
```

Comment:
Could you factor the two blocks in this commit out into a function?  I'm terrified of this bithackery getting out of sync.